odoo/addons/base/tests/test_res_users.py

```diff
@@ -217,3 +308,85 @@ def test_reified_groups_on_change(self):
+        """
+        with Form(self.test_group_user, view='base.view_users_form') as UserForm:
+            UserForm._values[self.field_service_cat_field] = self.group_field_service_user.id
+            UserForm._perform_onchange([self.field_service_cat_field])
+
+            self.assertFalse(UserForm.user_group_warning)
+
+    def test_user_group_inheritance_warning(self):
+        """
+            User makes changes to Sales User from Sales Administrator.
+            The warning should be there since Sales Administrator is Required
+            when User has Field Service Administrator
+        """
+        with Form(self.test_group_user, view='base.view_users_form') as UserForm:
+            UserForm._values[self.sales_cat_field] = self.group_sales_user.id
+            UserForm._perform_onchange([self.sales_cat_field])
+
+            self.assertEqual(
+                UserForm.user_group_warning,
+                'Since Test Group User is a/an Field Service Administrator, you can set Sales right only to Administrator'
+            )
+
+    def test_user_multi_group_inheritance_warning(self):
+        """
+           User makes changes To Sales User and Project User from Administrator
+           The warming should be there since Administrator for both are required
+           when User has Field Service Administrator
+        """
+        with Form(self.test_group_user, view='base.view_users_form') as UserForm:
+            UserForm._values[self.sales_cat_field] = self.group_sales_user.id
+            UserForm._values[self.project_cat_field] = self.group_project_user.id
+            UserForm._perform_onchange([self.sales_cat_field])
+
+            warnings = [
+                "Since Test Group User is a/an Field Service Administrator, you can set Sales right only to Administrator",
+                "Since Test Group User is a/an Field Service Administrator, you can set Project right only to Administrator"
+            ]
+            self.assertTrue(all(warning in UserForm.user_group_warning for warning in warnings))
+
+    def test_user_multi_possible_group_inheritance_warning(self):
+        """
+            User makes changes to Timesheets / User: Own Timesheet from Timesheets / Administrator
+            The warning should be there since Timessets / Administrator and Timesheets / User: All Documents
+            required when the User has Project Administrator
+        """
+        with Form(self.test_group_user, view='base.view_users_form') as UserForm:
+            UserForm._values[self.timesheet_cat_field] = self.group_timesheets_user_own_timesheet.id
+            UserForm._perform_onchange([self.timesheet_cat_field])
+
+            self.assertEqual(
+                UserForm.user_group_warning,
+                'Since Test Group User is a/an Project Administrator, you can set Timesheets right only to Administrator or User: all timesheets'
+            )
+
+    def test_user_group_empty_group_warning(self):
+        """
```

Comment:
the docstring doesn't follow the PEP
https://peps.python.org/pep-0257/

(tips: I use this to lint my docstring https://pypi.org/project/pydocstyle )